File: android/app/src/main/kotlin/com/goalmaker/app/data/settings/PostgrestProfileSettings.kt
package com.goalmaker.app.data.settings

import com.goalmaker.app.application.settings.ProfileSettings
import com.goalmaker.app.data.supabase.PostgrestHttp
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** [ProfileSettings] over PostgREST; row security limits the update to the owner's own profile. */
class PostgrestProfileSettings(private val postgrest: PostgrestHttp, private val userId: () -> String?) : ProfileSettings {
    override suspend fun update(timeZone: String, dayStartHour: Int) {
        val id = userId() ?: return
        postgrest.send(HttpMethod.Patch, "profiles") {
            parameter("id", "eq.$id")
            header("Prefer", "return=minimal")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("time_zone", timeZone); put("day_rollover_hour", dayStartHour) }.toString())
        }
    }
}

File: windows/src/GoalMaker.Infrastructure/Settings/PostgrestProfileSettings.cs
using System.Text;
using System.Text.Json.Nodes;
using GoalMaker.Core.Settings;
using GoalMaker.Infrastructure.Postgrest;

namespace GoalMaker.Infrastructure.Settings;

/// <summary><see cref="IProfileSettings"/> over PostgREST; row security limits the update to the owner's own profile.</summary>
public sealed class PostgrestProfileSettings(PostgrestHttp postgrest, Func<string?> userId) : IProfileSettings
{
    public async Task UpdateAsync(string timeZone, int dayStartHour, CancellationToken cancellationToken = default)
    {
        if (userId() is not { Length: > 0 } id)
        {
            return;
        }

        using var request = postgrest.Request(HttpMethod.Patch, $"profiles?id=eq.{Uri.EscapeDataString(id)}");
        request.Headers.Add("Prefer", "return=minimal");
        request.Content = new StringContent(
            new JsonObject { ["time_zone"] = timeZone, ["day_rollover_hour"] = dayStartHour }.ToJsonString(),
            Encoding.UTF8,
            "application/json");
        await postgrest.SendAsync(request, cancellationToken).ConfigureAwait(false);
    }
}
